### ISSUE TITLE
Discussion: Position utility

### ISSUE THREAD

**Issue by cooper-joe** [dhis2/ui-core#493]
Here is a suggestion for a `position` utility that can be used for popovers, tooltips and any other elements that are attached to an anchor.

### 12 available positions
![image](https://user-images.githubusercontent.com/33054985/66483170-d32d2c80-eaa4-11e9-8368-155b09e83b1d.png)

The default positions would also include an `offset` value that defines how far from the anchor it should be placed. `offset` could be customised (depending on how complex this is to do).

### Smart fitting
If an element, a popover for example, does not fit on screen in it's defined position:
  - Flip the orientation (top-left becomes bottom-left, right-bottom becomes left-bottom)

If said element still does not fit:
  - place in original position and resize, vertical or horizontal, to fit on screen.

If element still does not fit and cannot be resized:
  - place element offscreen. Layout/app has to handle this?

**Comment by Mohammer5** [comment]
> If an element, a popover for example, does not fit on screen in it's defined position:
> 
> Flip the orientation (top-left becomes bottom-left, right-bottom becomes left-bottom)
If said element still does not fit:
> 
> place in original position and resize, vertical or horizontal, to fit on screen.
If element still does not fit and cannot be resized:
> 
> place element offscreen. Layout/app has to handle this?

I think this is a good starting point.

<hr />

The current popover already has some logic that will adjust it's position, but in regards to the whole body, not the viewport. It's not resizing the element but moving it, [see the "Popover ( out of body )" stories"](https://deploy-preview-224--dhis2-ui-core.netlify.com/?path=/story/popover-out-of-body--bottom)

<hr />

Flipping positions can be very tricky, so this only would be ok with a limited amount of positions if we default to "render off screen" in case it doesn't fit into the screen.

My preferred solution would be to then be:
1. Check provided position
1. Check flipped position
1. Depends on component:
   * for the Popover:
     Show the contents in a modal
   * for the tooltip:
     Display the message a bit more prominently at the bottom/center of the viewport (similar to the alertbars)

No resizing or moving, using a Modal for the Popover will always work & using a message at the bottom of the screen for tooltips should also always work). Resizing & re-positioning is quite complicated, adds complexity (and therefore possible bugs) to the components and makes it a lot harder to maintain

**Comment by cooper-joe** [comment]
> My preferred solution would be to then be:
> 
>     1. Check provided position
> 
>     2. Check flipped position
> 
>     3. Depends on component:
>        
>        * for the Popover:
>          Show the contents in a modal
>        * for the tooltip:
>          Display the message a bit more prominently at the bottom/center of the viewport (similar to the alertbars)
> 
> 
> No resizing or moving, using a Modal for the Popover will always work & using a message at the bottom of the screen for tooltips should also always work). Resizing & re-positioning is quite complicated, adds complexity (and therefore possible bugs) to the components and makes it a lot harder to maintain

I recognize that this solve the issue of 'what to do with elements that don't fit', but I think this would be a confusing user experience. Content would present differently depending on scroll position, but how is a user to understand that? Also, `popover` vs. `modal` is not only a matter of visual presentaton. `Modal` is intended to interrupt a flow, is is therefore reserved for actions/info that needs to break a user flow and be dealt with. A `popover` is intended for the exact opposite, providing info inline without breaking flow. I think mixing between these based on screen size/position would result in a fragmented, difficult to understand user experience. The same applies to switching between a `tooltip` and an `alert bar`.

I think the better solution would be to display the element that doesn't fit offscreen. For example, a huge popover that doesn't fit on either axis of the body/viewport would have to get internal scrollbars. This is not ideal at all, but having this much content would also break our design and usage guidelines for the use of a popover, so it will never be ideal.

Of course, it is important to keep in mind we are discussing edge cases here. Most `tooltip`s, if used as intended, will fit on screen somewhere, for example. 

---

Note: I've updated my post above with a clearer example of the 12 positions.

**Comment by cooper-joe** [comment]
Here are some illustrated examples of each scenario and the suggested behaviour to aid discussion.

### Working as intended, space available
![image](https://user-images.githubusercontent.com/33054985/66806277-7065e600-ef27-11e9-9f29-db062cdccb3a.png)

### No space in intended position, using opposite position
![image](https://user-images.githubusercontent.com/33054985/66806343-98554980-ef27-11e9-965a-c38f5e333b57.png)

### No space in intended position, no space in opposite position, popover container not allowed to scroll. Body/app wrapper scrolls
![image](https://user-images.githubusercontent.com/33054985/66806394-b753db80-ef27-11e9-8589-35258b61f5f2.png)

### No space in intended position, no space in opposite position, popover container allowed to scroll
![image](https://user-images.githubusercontent.com/33054985/66806419-c6d32480-ef27-11e9-9d45-3b128748dfc3.png)

*The principles above would also apply to horizontal scrolling.

**Comment by ismay** [comment]
Is there a precedent for libraries flipping the intended position of a directional popup of there is no space? It seems counter intuitive to me for the user to specify a position, and then for us to not listen to that.

I've seen this kind of behaviour with dropdowns, but I don't think I've seen it happen when specifying a direction like you'd do with tooltips.

**Comment by cooper-joe** [comment]
> Is there a precedent for libraries flipping the intended position of a directional popup of there is no space? It seems counter intuitive to me for the user to specify a position, and then for us to not listen to that.
> 
> I've seen this kind of behaviour with dropdowns, but I don't think I've seen it happen when specifying a direction like you'd do with tooltips.

True, this does possibly display differently than the intended position, but in the pursuit of the main goal which is: display this thing on screen. If we don't flip the position then it won't be on screen. So I guess it depends how important position is, whether it's a preference or a hard requirement. 

I think allowing a `no-flip`/`no-reposition` or similar property would solve this for elements that *need* to be in a certain position. It would then be up to the implementer to decide 'is the position so important that it should be placed off-screen, even when other space/positions are available?'

**Comment by cooper-joe** [comment]
For reference, I have also worked on a project that used this library: https://popper.js.org/ , which also includes 'flip' functionality.

**Comment by ismay** [comment]
Yeah, I think it's important to signal clearly to the user that it might display in a different direction. A `noFlip` or `noReposition` toggle would be a clear way to indicate that. Or call the direction `preferredDirection`, something like that.

**Comment by cooper-joe** [comment]
Draft specs for components that need a position utility: https://github.com/dhis2/ui-core/pull/522
